src/config/src/Node.cpp

```diff
@@ -0,0 +1,382 @@
+        return _yaml;
+    }
+
+    Node::Type type() const {
+        return determineType(_yaml);
+    }
+
+    explicit operator bool() const {
+        return bool(_yaml);
+    }
+
+    bool isScalar() const {
+        return _yaml.IsScalar();
+    }
+
+    bool isMap() const {
+        return _yaml.IsMap();
+    }
+
+    bool isSequence() const {
+        return _yaml.IsSequence();
+    }
+
+    bool isNull() const {
+        return _yaml.IsNull();
+    }
+
+    size_t size() const {
+        return _yaml.size();
+    }
+
+    std::string key() const {
+        return _path.empty() ? "/" : _path.back().toString();
+    }
+
+    std::string path() const {
+        return joinPath(_path);
+    }
+
+    friend std::ostream& operator<<(std::ostream& out, const NodeImpl& impl) {
+        return out << YAML::Dump(impl._yaml);
+    }
+
+private:
+    friend NodeIterator;
+
+    static Children constructChildren(const v1::NodeKey::Path& path, ChildKeys& keyOrder, const YAML::Node yaml) {
+        Children out;
+        if (!yaml.IsMap() && !yaml.IsSequence()) {
+            return out;
+        }
+
+        int index = 0;
+        auto extractKey = [&](auto& kvp, YAML::Node iteratee) -> v1::NodeKey {
+            if (iteratee.IsMap()) {
+                return v1::NodeKey{kvp.first.template as<std::string>()};
+            } else {
+                return v1::NodeKey{index++};
+            }
+        };
+        auto extractValue = [&](auto& kvp, YAML::Node iteratee) -> YAML::Node {
+            if (iteratee.IsMap()) {
+                return kvp.second;
+            } else {
+                return kvp;
+            }
+        };
+
+        for (auto&& kvp : yaml) {
+            const auto key = extractKey(kvp, yaml);
+            keyOrder.push_back(key);
+            v1::NodeKey::Path childPath = path;
+            childPath.push_back(key);
+            out.emplace(key, std::make_unique<Node>(std::move(childPath), extractValue(kvp, yaml)));
+        }
+
+        return out;
+    }
+
+    static const YAML::Node _zombie;
+
+    const Node* _self;
+    // these 2 need to be mutable to generate placeholder nodes for non-existent keys.
+    mutable ChildKeys _keyOrder; // maintain insertion-order
+    mutable Children _children;
+    const YAML::Node _yaml;
+    const v1::NodeKey::Path _path;
+};
+
+const YAML::Node NodeImpl::_zombie = YAML::Load("")["zombie"];
+
+
+//
+// Node
+//
+
+Node::~Node() = default;
+
+Node::Node(const v1::NodeKey::Path& path, const YAML::Node yaml)
+    : _impl{std::make_unique<NodeImpl>(this, yaml, path)} {}
+
+std::string Node::key() const {
+    return this->_impl->key();
+}
+
+std::string Node::path() const {
+    return this->_impl->path();
+}
+
+std::ostream& operator<<(std::ostream& out, const Node& node) {
+    return out << *(node._impl);
+}
+
+bool Node::isScalar() const {
+    return _impl->isScalar();
+}
+
+Node::Type Node::type() const {
+    return _impl->type();
+}
+
+bool Node::isNull() const {
```

Comment:
Maybe not the right thing to do but if you just check the _impl's type directly, you won't have to write the `isXXXX` functions on the impl again.

Reply:
I considered that but I wanted the `class Node` functions to basically be "dumb" to keep all the logic in `NodeImpl`. It turned out better that way when I added "strict mode".